syntax = "proto3";
package natun.core.v1alpha1;
option go_package = "github.com/natun-ai/natun/core/v1alpha1;coreV1alpha1";

import "validate/validate.proto";
import "google/protobuf/timestamp.proto";
import "google/api/annotations.proto";
import "natun/core/v1alpha1/types.proto";

// +++ Read methods +++

// HistoricalGetRequest
message HistoricalGetRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  google.protobuf.Timestamp start_time = 3;
  google.protobuf.Timestamp end_time = 4;
}
message HistoricalGetResponse {
  string uuid = 1 [(validate.rules).string.uuid = true];
  repeated FeatureValue values = 2;
}

// GetRequest
message GetRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  string entity_id = 3;
}
message GetResponse {
  string uuid = 1 [(validate.rules).string.uuid = true];
  FeatureValue value = 2;
  Metadata metadata = 3;
}

// +++ Write methods +++

message SetRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  string entity_id = 3;
  Value value = 4;
  google.protobuf.Timestamp timestamp = 5;
}
message SetResponse {
    string uuid = 1 [(validate.rules).string.uuid = true];
    google.protobuf.Timestamp timestamp = 2;
}

message AppendRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  string entity_id = 3;
  Scalar value = 4;
  google.protobuf.Timestamp timestamp = 5;
}
message AppendResponse {
    string uuid = 1 [(validate.rules).string.uuid = true];
    google.protobuf.Timestamp timestamp = 2;
}

message IncrRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  string entity_id = 3;
  Scalar value = 4;
  google.protobuf.Timestamp timestamp = 5;
}
message IncrResponse {
    string uuid = 1 [(validate.rules).string.uuid = true];
    google.protobuf.Timestamp timestamp = 2;
}

message UpdateRequest {
  string uuid = 1 [(validate.rules).string.uuid = true];
  string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
  string entity_id = 3;
  Value value = 4;
  google.protobuf.Timestamp timestamp = 5;
}
message UpdateResponse {
    string uuid = 1 [(validate.rules).string.uuid = true];
    google.protobuf.Timestamp timestamp = 2;
}
message MetadataRequest {
    string uuid = 1 [(validate.rules).string.uuid = true];
    string fqn = 2 [(validate.rules).string.pattern = "(i?)^([a0-z9\\-\\.]*)(\\[([a0-z9])*\\])?$"];
}
message MetadataResponse {
    string uuid = 1 [(validate.rules).string.uuid = true];
    Metadata metadata = 2;
}


// +++ Service +++
service EngineService {
  rpc Metadata(MetadataRequest) returns (MetadataResponse) {
    option (google.api.http) = {
      get: "/{fqn}"
    };
  }
  rpc Get(GetRequest) returns (GetResponse) {
    option (google.api.http) = {
      get: "/{fqn}/{entity_id}"
    };
  }
  rpc HistoricalGet(HistoricalGetRequest) returns (HistoricalGetResponse) {
    option (google.api.http) = {
      post: "/{fqn}/historical"
      body: "*"
    };
  }
  rpc Set(SetRequest) returns (SetResponse) {
    option (google.api.http) = {
      put: "/{fqn}/{entity_id}"
    };
  }
  rpc Append(AppendRequest) returns (AppendResponse) {
    option (google.api.http) = {
      post: "/{fqn}/{entity_id}/append"
    };
  }
  rpc Incr(IncrRequest) returns (IncrResponse) {
    option (google.api.http) = {
      post: "/{fqn}/{entity_id}/incr"
    };
  }
  rpc Update(UpdateRequest) returns (UpdateResponse) {
    option (google.api.http) = {
      post: "/{fqn}/{entity_id}"
    };
  }
}